subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+
+[subs="attributes"]
+[source,kotlin]
+----
+dependencies {
+    compileOnly("dev.gradleplugins:gradle-api:{gradleVersion}")
+}
+----
+
+TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.
+
+This makes the version of the Gradle API being used to build the plugin more explicit, removing the ambiguity avoids the implicit link between `gradleApi()` and the build version of Gradle.
+
+On the other hand, you cannot use newer APIs if your plugin is running on a recent version of Gradle.
+
+[[sec:gradle_kotlin_dsl]]
+=== gradleKotlinDsl()
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly(gradleKotlinDsl())
+}
+----
+
+The `gradleKotlinDsl()` contains helpers to work better with Kotlin.
+
+For an example:
+
+* `tasks.withType<Jar> {}` (instead of `tasks.withType(Jar::class.java) {}`)
+* `val commonMain by getting {}` (instead of `getByName("commonMain") {}`)
+* `"implementation"("group:artifact:version")` (instead of `add("implementation", "group:artifact:version")`)
+* Even `gradleKotlinDsl()` is provided by `gradleKotlinDsl()` !
+
+You can use optionally it to build your plugins, but it's not mandatory.
+
+It is part of the Gradle distribution. You will find it under `~/.gradle/wrapper/dists/gradle-$version-bin/$sha/gradle-$version/lib/gradle-kotlin-dsl-$version.jar`.
+
+When writing `build.gradle.kts`, `gradleKotlinDsl()` is in the classpath and provide helpers to work better with Kotlin.
+
+TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.
+
+[[sec:kotlin_stdlib]]
+=== kotlin-stdlib
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-stdlib:$version")
+}
+----
+
+The Kotlin compatibility matrix can be found in the https://docs.gradle.org/current/userguide/compatibility.html#kotlin[Gradle compatibility page]. If you want your plugin to be compatible with a given version of Gradle, you need to make sure to use a compatible version of kotlin-stdlib. For an example, if you need your plugin to be compatible with Gradle 7.0, you cannot use any Kotlin 1.5 only API (like https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/lowercase.htmllowercase) and need to compile against `kotlin-stdlib:1.4.31`.
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-stdlib:1.4.31")
+}
+----
+
+For similar reasons, any transitive dependency of your plugin needs to use a `kotlin-stdlib` version compatible with the Gradle one. While you can ensure that at compile time for your own code, it's often more complicated for dependencies so make sure to test your plugins and/or double check your transitive dependencies.
+
+Note: The `kotlin-stdlib` version is independent of your Kotlin Gradle Plugin version. You can use a newer Kotlin Stdlib version with an older `kotlin-stdlib` using the `apiVersion` compiler argument.
```

Comment:
The last sentence is a bit confusing. Did you mean Kotlin Gradle Plugin instead?

Reply:
Also, you can put a link to apiVersion section below.

Reply:
> The last sentence is a bit confusing. Did you mean Kotlin Gradle Plugin instead?

Completely! Thanks!

> Also, you can put a link to apiVersion section below.

Good idea 👍